_posts/24-07-2022-what_is_a_reader_and_why_it_matters.md

```diff
@@ -0,0 +1,134 @@
+---
+layout: post
+title: A primer on Readers
+category: Scala
+description: Without mentioning monads!
+---
+### The dream
+Imagine we have functions:
+
+```Scala
+def f: A => B = ???
+def g: B => C = ???
+def h: C => D = ???
+```
+
+The implementation may be whatever we want it to be, these functions might parse some data, perform calculations, format strings et cetera. Now, often we want to chain multiple operations together - ie. parse a JSON with sales data, then extract monthly sales figures and finally output an average. After all, coding is all about gluing blocks together to create a useful program.
+We could, of course, write `h(g(f(x)))` each time we want the whole functionality, but if it's used frequently, giving it a proper name is a good idea.
+Scala gives us many different ways of gluing functions together, but I'm going to stick to the `andThen` from `Function1` trait - because it behaves almost like the mathematical `∘` operator, just in reverse order. With it creating a function `i: A => D` is as simple as writing:
```

Comment:
For clarity, I'd mention that there's `compose` that behaves exactly like `∘`... it's just much harder to follow than `andThen` ;p

Reply:
Added a footnote.